File: components_cpp/src/manual_composition.cpp
#include "rclcpp/rclcpp.hpp"
#include "components_cpp/node1.hpp"
#include "components_cpp/node2.hpp"

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    // Criar um executor single-threaded
    rclcpp::executors::SingleThreadedExecutor executor;

    // Criar instâncias dos nodes
    auto node1 = std::make_shared<Node1>();
    auto node2 = std::make_shared<Node2>();

    // Adicionar os nodes ao executor
    executor.add_node(node1);
    executor.add_node(node2);

    // Executar o executor
    executor.spin();

    rclcpp::shutdown();
    return 0;
}

File: components_py/components_py/manual_composition.py
#!/usr/bin/env python3
import rclpy
from rclpy.executors import SingleThreadedExecutor
from components_py.node1 import Node1
from components_py.node2 import Node2

"""Este exemplo demonstra a composição manual de múltiplos nós em um único executor.
Usamos o SingleThreadedExecutor para gerenciar a execução do Node1 e Node2.
Cada nó possui múltiplos timers que simulam callbacks demorados usando time.sleep().
Ao executar este script, você verá que os callbacks dos timers são chamados de forma sequencial,
devido à natureza do SingleThreadedExecutor.
"""

def main(args=None):
    rclpy.init(args=args)

    node1 = Node1()
    node2 = Node2()

    executor = SingleThreadedExecutor()
    executor.add_node(node1)
    executor.add_node(node2)
    executor.spin()

    rclpy.shutdown()

if __name__ == "__main__":
    main()